### PR TITLE
URL blacklist

### LINKED ISSUE #219: Stop following canonical urls on certain conditions
Is there some clever way to stop following canonical URLs on certain conditions?

I see lots of different webpages pointing Embed to the front page, when I've actually requested some other page (that also exists). This happens all the time, and most of the time, it puts Facebook as referrer in the url. This isn't Embed's fault, as it's often some canonical-like meta-tag somewhere pointing to that specific URL.

**Example:** [http://glomdalen.no/**skarnes**](http://glomdalen.no/skarnes)
**Redirects to:** [**https://www**.glomdalen.no/**skarnes**](https://www.glomdalen.no/skarnes)
**While `og:url` points to:** [**http://www.glomdalen.no**?ns_campaign=frontpage&amp;ns_mchannel=recommend_button&amp;ns_source=facebook&amp;ns_linkname=facebook&amp;ns_fee=0](http://www.glomdalen.no?ns_campaign=frontpage&amp;ns_mchannel=recommend_button&amp;ns_source=facebook&amp;ns_linkname=facebook&amp;ns_fee=0)
Witch becomes the final URL according to Embed (a different page).
Visiting that in a browser, causes redirection to HTTPS (once again)

How about not following so-called `canonical` urls if it contains `campaign` or `source` parameters? We'll already received a HTTP 200 right, so why bother adding that referrer, witch _may or may not_ point us to a potentially different and unwanted page...

Parameter blacklist suggestion:
- `ns_campaign`
- `ns_source`
- `utm_campaign`
- `utm_medium`
- `utm_source`

_Offtopic: Dealing with the webpages, requesting them to fix it isn't the right way to go either, it's too many of them, and often, this is how they track facebook clicks..._

EDIT: typo + bold text

### PR DESCRIPTION
**Pull request by JanPetterMG**
- URL blacklist config array
- Some cosmetic fixes (PHPStorm complained a lot about missing return types and values)
- Readme updated (including fixing the broken table)

Fixes #219

### CODE REVIEW

**Comment by JanPetterMG** [comment]
Now all urls containing for example `utm_source=` will fail, but it should also check for `?` and `&` prefixes, witch it currently doesn't.

This regex works, but can't get it working in Embed...
https://regex101.com/r/4pPfpv/2

Except for that, this PR is ready for review/merge

File: src/Embed.php [hunk 1]
```diff
@@ -3,27 +3,67 @@
...
      */
-    public static function create($url, array $config = [], DispatcherInterface $dispatcher = null)
+    public static function create($url, array $config = null, DispatcherInterface $dispatcher = null)
     {
         if (!($url instanceof Url)) {
             $url = Url::create($url);
         }
 
+        if ($config === null) {
+            $dispatcher = self::$default_config;
```
Comment (oscarotero) [inline src/Embed.php]: You're assigning the $default_config to the wrong variable.

Comment (JanPetterMG) [inline src/Embed.php]: Ups, didn't see that one, nice catch. Was copy pasting the next 3 lines, but forgot to change...
Fixed now.

File: src/Http/Url.php [hunk 2]
```diff
@@ -68,9 +68,8 @@ public function match($patterns)
             $patterns = [$patterns];
         }
 
-        //Remove scheme and query
-        $url = preg_replace('|(\?.*)?$|', '', (string) $this);
-        $url = preg_replace('|^(\w+://)|', '', $url);
+        //Remove scheme
+        $url = preg_replace('|^(\w+://)|', '', (string) $this);
```
Comment (JanPetterMG) [inline src/Http/Url.php]: I know, but it's also essential to be able to block urls with specific parameters, witch is the main goal with this PR.
The obvious alternative is to re-invent `match()`, and specifying an blacklist for each part of the url (scheme, host, port, parameter, query).

The definite best solution would be switching over to unconditional regex, letting the regex string specify every requirement, but again, that may also have unintended consequences for the minority.

How about double configuration blacklists then? and deprecate the old one?
- `images_blacklist` and `image_blacklist_regex`
- `url_blacklist_regex`

Comment (oscarotero) [inline src/Http/Url.php]: I've changed `match()` to support query parameteres, but it must be specified explicity using the `?` character, as explained here: https://github.com/oscarotero/Embed/pull/221#pullrequestreview-44115897

Can you update the pull request with this change and then update the black list patterns to include the `?`?

Comment (JanPetterMG) [inline src/Http/Url.php]: Nice workaround, but it also makes it impossible to require the parameter to start with either `?` or `&` in a single string. I've fixed that by "doubling" the blacklist, one string for each.

Comment (oscarotero) [inline src/Http/Url.php]: I see. I've added a bit more magic. If the pattern contains `?&`, is replaced to `[\?&]`, so you could do simply `?&utm_medium=*`

Comment (JanPetterMG) [inline src/Http/Url.php]: Now it works!

**Comment by oscarotero** [comment]
Nice work. Thank you.

**Pull request merged**